docs/sc-memory/api/cpp/extended/agents/agents.md

```diff
@@ -0,0 +1,690 @@
+{
+public:
+  ScAddr GetActionClass() const override;
+  ScResult DoProgram(ScAction & action) override;
+
+  // Other user-defined methods.
+};
+```
+
+See [**C++ Events API**](events.md) and [**C++ Event subscriptions API**](event_subscriptions.md) to learn how to use and handle sc-events.
+
+!!! note
+    Define action class as keynode in agent class or keynodes class.
+
+!!! warning
+    You must override methods `GetActionClass` and `DoProgram`.
+
+You can specify any existing event types as a template argument to the `ScAgent` class. For example, you can create agent that will be triggered to sc-event of removing sc-element.
+
+```cpp
+// File my_agent.hpp
+#pragma once
+
+#include <sc-memory/sc_agent.hpp>
+
+class MyAgent : public ScAgent<ScEventBeforeEraseElement>
+{
+public:
+  ScAddr GetActionClass() const override;
+  ScResult DoProgram(
+    ScEventBeforeEraseElement const & event, ScAction & action) override;
+
+  // Other user-defined methods.
+};
+```
+
+If you want to change specification of this agent in knowledge base, then write like this:
+
+```cpp
+// File my_agent.hpp
+#pragma once
+
+#include <sc-memory/sc_agent.hpp>
+
+class MyAgent : public ScBaseAgent
+{
+public:
+  ScAddr GetActionClass() const override;
+  ScResult DoProgram(
+    ScElementaryEvent const & event, ScAction & action) override;
+
+  // Other user-defined methods.
+};
+```
+
+This implementation allows to provide any sc-event type to `DoProgram`.
+
+!!! note
+    `ScBaseAgent` is alias for `ScAgent<ScElementaryEvent>`.
+
+!!! note
+    If you provide specification of your agent in knowledge base, then you don't need to override `GetActionClass`. See [**C++ Modules API**](modules.md) to learn how to implement agents with specification in the knowledge base.
+
+### **ScActionInitiatedAgent**
+
+In multi-agent systems most of the agents are implemented to execute actions initiated by other agents. While `ScAgent` is useful to create broad event handling logic, using it to handle action initiations requires some boilerplate. We've implemented another agent class to make it easier for our users to implement action-executing agents. These agents are named action agents. Implementing these agents requires passing action class node rather than checking initiation condition manually.
+
+This class can be only used for agents that should be triggered by generating an output sc-arc from `action_initiated` class node.
+
+```cpp
+// File my_agent.hpp
+#pragma once
+
+#include <sc-memory/sc_agent.hpp>
+
+// Inherit your agent class from `ScActionInitiatedAgent` class.
+class MyAgent : public ScActionInitiatedAgent
+{
+public:
+  // Here you should specify class of actions which the given agent performs. 
+  // This overriding is required.
+  ScAddr GetActionClass() const override;
+  // Here you should implement program of the given agent. 
+  // This overriding is required.
+  ScResult DoProgram(ScActionEvent const & event, ScAction & action) override;
```

Comment:
shouldn't it now be called ScActionInitiatedEvent?

Reply:
fixed